httpload/cmd/httpload/main.go

```diff
@@ -0,0 +1,440 @@
+	}
+	defer cancel()
+
+	go func() {
+		<-done
+		cancel()
+	}()
+
+	var errcList []<-chan error
+	urlc, errc, err := config.genRequests(ctx)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	resultc, errc, err := lt.genLoadRequests(ctx, config, urlc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	errc, err = lt.genResults(ctx, resultc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	err = lt.waitForPipeline(errcList...)
+
+	done <- true
+
+	return err
+}
+
+func (lt *LoadTester) dumpStatuses(w io.Writer) {
+	statuses := make(map[int]int)
+
+	for _, result := range lt.results {
+		statuses[result.status]++
+	}
+
+	keys := make([]int, 0, len(statuses))
+
+	for key := range statuses {
+		keys = append(keys, key)
+	}
+
+	sort.Ints(keys)
+
+	for _, key := range keys {
+		code := float64(key)
+		count := float64(statuses[key])
+		fmt.Fprintf(w, "%s\n", fmt.Sprintf("[%3.f] %12.f response(s)", code, count))
+	}
+}
+
+func (lt *LoadTester) dumpTimings(w io.Writer) {
+	totalTiming := float64(0)
+	maxTiming := float64(0)
+	minTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(lt.results)
+	noResults := (n == 0)
+
+	if noResults {
+		return
+	}
+
+	timing0 := lt.results[0].timing.Seconds()
+	totalTiming = timing0
+	maxTiming = timing0
+	minTiming = timing0
+
+	for i := 1; i < n; i++ {
+		timing := lt.results[i].timing.Seconds()
+
+		totalTiming += timing
+
+		if timing > maxTiming {
+			maxTiming = timing
+		}
+
+		if timing < minTiming {
+			minTiming = timing
+		}
+	}
+
+	averageTiming = totalTiming / float64(n)
+	requestsPerSecond = float64(1) / averageTiming
+
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Total:        %12.4f secs", totalTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", maxTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", minTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+}
+
+func setupGracefulShutdown(done chan bool) {
+	signalChannel := make(chan os.Signal, 1)
+
+	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGKILL)
+
+	<-signalChannel
+
+	done <- true
+}
+
+func trueMain(flags *flag.FlagSet, args []string) int {
+	done := make(chan bool)
+	defer close(done)
+	go setupGracefulShutdown(done)
+
+	config := newConfigFromFlags(flags)
+
+	if err := config.parse(flags, args); err != nil {
+		fmt.Println("main: error during arguments parsing. Error:", err)
+		flags.Usage()
+		return exitCodeError
+	}
+
+	if err := config.checkFlags(); err != nil {
+		log.Println("main: error checking flags. Error:", err)
+		return exitCodeError
+	}
+
+	loadTester := newLoadTesterFromConfig(config)
+
+	go func() {
+		err := loadTester.run(config, done)
+
+		if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
`loadTester.run()` is leaking its implementation details here (the fact that `run` internally uses a context).
IMHO `run` should have itself deal with `context.DeadlineExceeded` and not require `trueMain` to know about it (i.e returns a nil error since it's not really an error)

Reply:
How do you suggest to solve the problem? Do I have to write a `run`/`trueRun` couple of methods? The alternative is to check for `DeadlineExceeded` on all `run`'s `ifs.

Reply:
Do not mind it, maybe I solved.